### ISSUE TITLE
[nvmf_timeout] Failure while checking reconnect counter

### ISSUE THREAD

**Issue by mikeBashStuff** [spdk/spdk#2595]
https://ci.spdk.io/results/autotest-per-patch/builds/80709/archive/nvmf-tcp-vg-autotest/build.log

Point of failure:
```
00:19:07.219  ========== Backtrace start: ==========
00:19:07.219  
00:19:07.219  in test/nvmf/host/timeout.sh:137 -> main(["--transport=tcp"])
00:19:07.219       ...
00:19:07.219     132 	sleep 5
00:19:07.219     133 	cat $testdir/trace.txt
00:19:07.219     134 	
00:19:07.219     135 	# Check the frequency of delay reconnect
00:19:07.219     136 	if (("$(grep -c "reconnect delay bdev controller NVMe0" < $testdir/trace.txt)" != 2)); then
00:19:07.219  => 137 		false
00:19:07.219     138 	fi
00:19:07.219     139 	
00:19:07.219     140 	kill $dtrace_pid
00:19:07.219     141 	rm -f $testdir/trace.txt
00:19:07.219     142 	wait $rpc_pid
00:19:07.219       ...
00:19:07.219  
00:19:07.219  ========== Backtrace end ==========
```

**Comment by spdkci** [comment]
Another instance of this failure. Reported by @mikeBashStuff. log: https://ci.spdk.io/public_build/autotest-per-patch_80709.html

**Comment by jimharris** [comment]
> https://ci.spdk.io/results/autotest-per-patch/builds/80709/archive/nvmf-tcp-vg-autotest/build.log
> 
> Point of failure:
> 
> ```
> 00:19:07.219  ========== Backtrace start: ==========
> 00:19:07.219  
> 00:19:07.219  in test/nvmf/host/timeout.sh:137 -> main(["--transport=tcp"])
> 00:19:07.219       ...
> 00:19:07.219     132 	sleep 5
> 00:19:07.219     133 	cat $testdir/trace.txt
> 00:19:07.219     134 	
> 00:19:07.219     135 	# Check the frequency of delay reconnect
> 00:19:07.219     136 	if (("$(grep -c "reconnect delay bdev controller NVMe0" < $testdir/trace.txt)" != 2)); then

This part of the log is also important - it shows that we actually found no instances of this string in the log:

```
00:19:07.218   12:10:06	-- host/timeout.sh@133 -- # cat /home/vagrant/spdk_repo/spdk/test/nvmf/host/trace.txt
00:19:07.218  Attaching 5 probes...
00:19:07.218    12:10:06	-- host/timeout.sh@136 -- # grep -c 'reconnect delay bdev controller NVMe0'
00:19:07.218     12:10:06	-- host/timeout.sh@136 -- # trap - ERR
00:19:07.218     12:10:06	-- host/timeout.sh@136 -- # print_backtrace
00:19:07.218     12:10:06	-- common/autotest_common.sh@998 -- # [[ hxBET =~ e ]]
00:19:07.218     12:10:06	-- common/autotest_common.sh@998 -- # return 0
00:19:07.218   12:10:06	-- host/timeout.sh@136 -- # (( 0 != 2 ))
00:19:07.218   12:10:06	-- host/timeout.sh@137 -- # false
```


> 00:19:07.219  => 137 		false
> 00:19:07.219     138 	fi
> 00:19:07.219     139 	
> 00:19:07.219     140 	kill $dtrace_pid
> 00:19:07.219     141 	rm -f $testdir/trace.txt
> 00:19:07.219     142 	wait $rpc_pid
> 00:19:07.219       ...
> 00:19:07.219  
> 00:19:07.219  ========== Backtrace end ==========
> ```

**Comment by spdkci** [comment]
Another instance of this failure. Reported by @MengjinWu. log: https://ci.spdk.io/public_build/autotest-per-patch_80776.html

**Comment by spdkci** [comment]
Another instance of this failure. Reported by @mikeBashStuff. log: https://ci.spdk.io/public_build/autotest-per-patch_80772.html

**Comment by spdkci** [comment]
Another instance of this failure. Reported by @peluse. log: https://ci.spdk.io/public_build/autotest-per-patch_80948.html

**Comment by spdkci** [comment]
Another instance of this failure. Reported by @mikeBashStuff. log: https://ci.spdk.io/public_build/autotest-per-patch_81025.html

**Comment by spdkci** [comment]
Another instance of this failure. Reported by @EugeneKochetov. log: https://ci.spdk.io/public_build/autotest-per-patch_81052.html

**Comment by spdkci** [comment]
Another instance of this failure. Reported by @mikeBashStuff. log: https://ci.spdk.io/public_build/autotest-per-patch_81079.html

**Comment by spdkci** [comment]
Another instance of this failure. Reported by @shuhei-matsumoto. log: https://ci.spdk.io/public_build/autotest-per-patch_81082.html

**Comment by spdkci** [comment]
Another instance of this failure. Reported by @mikeBashStuff. log: https://ci.spdk.io/public_build/autotest-per-patch_81236.html

**Comment by spdkci** [comment]
Another instance of this failure. Reported by @peluse. log: https://ci.spdk.io/public_build/autotest-per-patch_81390.html

**Comment by spdkci** [comment]
Another instance of this failure. Reported by @peluse. log: https://ci.spdk.io/public_build/autotest-per-patch_81389.html

**Comment by spdkci** [comment]
Another instance of this failure. Reported by @peluse. log: https://ci.spdk.io/public_build/autotest-per-patch_81391.html

**Comment by spdkci** [comment]
Another instance of this failure. Reported by @peluse. log: https://ci.spdk.io/public_build/autotest-per-patch_81403.html

**Comment by spdkci** [comment]
Another instance of this failure. Reported by @peluse. log: https://ci.spdk.io/public_build/autotest-per-patch_81393.html

**Comment by spdkci** [comment]
Another instance of this failure. Reported by @peluse. log: https://ci.spdk.io/public_build/autotest-per-patch_81414.html

**Comment by spdkci** [comment]
Another instance of this failure. Reported by @peluse. log: https://ci.spdk.io/public_build/autotest-per-patch_81506.html

**Comment by changpe1** [comment]
I added a patch to print error information when got this error via https://ci.spdk.io/results/autotest-per-patch/builds/81532/archive/nvmf-tcp-vg-autotest/build.log, and found that indeed there are no error IOs reported in client when removing listener from target, so the bdev nvme will not enter expected error reconnect.

**Comment by spdkci** [comment]
Another instance of this failure. Reported by @mmkayPL. log: https://ci.spdk.io/public_build/autotest-per-patch_81568.html

**Comment by spdkci** [comment]
Another instance of this failure. Reported by @jimharris. log: https://ci.spdk.io/public_build/autotest-per-patch_81607.html

**Comment by spdkci** [comment]
Another instance of this failure. Reported by @jimharris. log: https://ci.spdk.io/public_build/autotest-per-patch_81608.html

**Comment by spdkci** [comment]
Another instance of this failure. Reported by @optimistyzy. log: https://ci.spdk.io/public_build/autotest-per-patch_81565.html

**Comment by spdkci** [comment]
Another instance of this failure. Reported by @mmkayPL. log: https://ci.spdk.io/public_build/autotest-per-patch_81611.html

**Comment by spdkci** [comment]
Another instance of this failure. Reported by @mmkayPL. log: https://ci.spdk.io/public_build/autotest-per-patch_81604.html

**Comment by changpe1** [comment]
I added one patch https://review.spdk.io/gerrit/c/spdk/spdk/+/13805 to improve this test, run 7 CI tests and didn't see the same issue with this patch.

**Comment by spdkci** [comment]
Another instance of this failure. Reported by @mmkayPL. log: https://ci.spdk.io/public_build/autotest-per-patch_81661.html

**Comment by spdkci** [comment]
Another instance of this failure. Reported by @mmkayPL. log: https://ci.spdk.io/public_build/autotest-per-patch_81667.html

**Comment by spdkci** [comment]
Another instance of this failure. Reported by @mikeBashStuff. log: https://ci.spdk.io/public_build/autotest-per-patch_81677.html

**Comment by spdkci** [comment]
Another instance of this failure. Reported by @mikeBashStuff. log: https://ci.spdk.io/public_build/autotest-per-patch_81680.html

**Comment by jimharris** [comment]
@changpe1 is actively debugging this issue, so assigning to him.

**Comment by spdkci** [comment]
Another instance of this failure. Reported by @shuhei-matsumoto. log: https://ci.spdk.io/public_build/autotest-per-patch_81606.html

**Comment by spdkci** [comment]
Another instance of this failure. Reported by @changpe1. log: https://ci.spdk.io/public_build/autotest-per-patch_81722.html

**Comment by changpe1** [comment]
The issue may still happen after my patch to relax the failure condition.

Think more for the test case, I think https://review.spdk.io/gerrit/c/spdk/spdk/+/13840 should fix the issue, the client expects a socket READ error, this will not happen for NVMe WRITE commands.

**Comment by spdkci** [comment]
Another instance of this failure. Reported by @yidong0635. log: https://ci.spdk.io/public_build/autotest-per-patch_81726.html

**Comment by spdkci** [comment]
Another instance of this failure. Reported by @Comphix. log: https://ci.spdk.io/public_build/autotest-per-patch_81846.html

**Comment by spdkci** [comment]
Another instance of this failure. Reported by @Comphix. log: https://ci.spdk.io/public_build/autotest-per-patch_81850.html

**Comment by changpe1** [comment]
Looked all the failure cases, all failures happened in `VM-host-PE1`, the same test case also run in `VM-host-SM16`, the only difference between the two test environment is socket interface, one is using POSIX interface while another one is using URING.

**Comment by spdkci** [comment]
Another instance of this failure. Reported by @Comphix. log: https://ci.spdk.io/public_build/autotest-per-patch_81851.html

**Comment by spdkci** [comment]
Another instance of this failure. Reported by @yidong0635. log: https://ci.spdk.io/public_build/autotest-per-patch_81863.html

**Comment by spdkci** [comment]
Another instance of this failure. Reported by @ksztyber. log: https://ci.spdk.io/public_build/autotest-per-patch_81889.html

**Comment by spdkci** [comment]
Another instance of this failure. Reported by @Comphix. log: https://ci.spdk.io/public_build/autotest-per-patch_81862.html

**Comment by spdkci** [comment]
Another instance of this failure. Reported by @ksztyber. log: https://ci.spdk.io/public_build/autotest-per-patch_81898.html

**Comment by spdkci** [comment]
Another instance of this failure. Reported by @ksztyber. log: https://ci.spdk.io/public_build/autotest-per-patch_81890.html

**Comment by spdkci** [comment]
Another instance of this failure. Reported by @ksztyber. log: https://ci.spdk.io/public_build/autotest-per-patch_81912.html

**Comment by spdkci** [comment]
Another instance of this failure. Reported by @peluse. log: https://ci.spdk.io/public_build/autotest-per-patch_81816.html

**Comment by spdkci** [comment]
Another instance of this failure. Reported by @peluse. log: https://ci.spdk.io/public_build/autotest-per-patch_81965.html

**Comment by spdkci** [comment]
Another instance of this failure. Reported by @peluse. log: https://ci.spdk.io/public_build/autotest-per-patch_81983.html

**Comment by spdkci** [comment]
Another instance of this failure. Reported by @mikeBashStuff. log: https://ci.spdk.io/public_build/autotest-per-patch_82040.html

**Comment by shuhei-matsumoto** [comment]
Hi, If I see correctly, this issue is not fixed by Changpeng's patch yet. I am checking why reconnect timer was not expired even if reconnect succeeded.

**Comment by shuhei-matsumoto** [comment]
It is likely that my first comment was wrong but I am checking the timeout test in detail anyway.

**Comment by shuhei-matsumoto** [comment]
https://ci.spdk.io/results/autotest-per-patch/builds/82025/archive/nvmf-tcp-vg-autotest/build.log reproduced this failure. It looks that even randread did not cause error.

**Comment by spdkci** [comment]
Another instance of this failure. Reported by @jimharris. log: https://ci.spdk.io/public_build/autotest-per-patch_81990.html

**Comment by spdkci** [comment]
Another instance of this failure. Reported by @mikeBashStuff. log: https://ci.spdk.io/public_build/autotest-per-patch_82101.html

**Comment by spdkci** [comment]
Another instance of this failure. Reported by @mikeBashStuff. log: https://ci.spdk.io/public_build/autotest-per-patch_82105.html

**Comment by spdkci** [comment]
Another instance of this failure. Reported by @mikeBashStuff. log: https://ci.spdk.io/public_build/autotest-per-patch_82116.html

**Comment by spdkci** [comment]
Another instance of this failure. Reported by @mikeBashStuff. log: https://ci.spdk.io/public_build/autotest-per-patch_82113.html

**Comment by spdkci** [comment]
Another instance of this failure. Reported by @mikeBashStuff. log: https://ci.spdk.io/public_build/autotest-per-patch_82122.html

**Comment by shuhei-matsumoto** [comment]
I created a patch https://review.spdk.io/gerrit/c/spdk/spdk/+/13908
I guess the default timeout is too long to detect connection failure within the test interval.
If this idea is wrong, I will continue investigation too.

**Comment by spdkci** [comment]
Another instance of this failure. Reported by @mikeBashStuff. log: https://ci.spdk.io/public_build/autotest-per-patch_82132.html

**Comment by mikeBashStuff** [comment]
> I created a patch https://review.spdk.io/gerrit/c/spdk/spdk/+/13908
I guess the default timeout is too long to detect connection failure within the test interval.
If this idea is wrong, I will continue investigation too.

Unfortunately, this patch didn't help - I've run nvmf_timeout in a loop for quite some time with this patch on top and the failure rate was similar to what is currently being seen on master.

**Comment by spdkci** [comment]
Another instance of this failure. Reported by @mikeBashStuff. log: https://ci.spdk.io/public_build/autotest-per-patch_82142.html

**Comment by shuhei-matsumoto** [comment]
Hi @mikeBashStuff Thank you so much for your repeated testing. CI passed once but it would be rare as you already tested. I hope you find the root cause and fix it.

**Comment by changpe1** [comment]
> Hi @mikeBashStuff Thank you so much for your repeated testing. CI passed once but it would be rare as you already tested. I hope you find the root cause and fix it.

When got failure test result, the client log is also printed now, there aren't any connection errors and IO errors happened in the client from the log, I am thinking it's related with VM network environment, because I can't reproduce this issue in my local environment, @mikeBashStuff what's the environment do you use to reproduce this issue?

**Comment by changpe1** [comment]
> > Hi @mikeBashStuff Thank you so much for your repeated testing. CI passed once but it would be rare as you already tested. I hope you find the root cause and fix it.
> 
> When got failure test result, the client log is also printed now, there aren't any connection errors and IO errors happened in the client from the log, I am thinking it's related with VM network environment, because I can't reproduce this issue in my local environment, @mikeBashStuff what's the environment do you use to reproduce this issue?

The same test is running in two VM environments, one is URING, the failed one is POSIX, we can only see this issue in the POSIX one.

**Comment by spdkci** [comment]
Another instance of this failure. Reported by @yidong0635. log: Not Found

**Comment by spdkci** [comment]
Another instance of this failure. Reported by @AlekseyMarchuk. log: https://ci.spdk.io/public_build/autotest-per-patch_82139.html

**Comment by shuhei-matsumoto** [comment]
Hi @mikeBashStuff @changpe1 

The patch https://review.spdk.io/gerrit/c/spdk/spdk/+/13908 passed CI 3 times successively.
I believe this is not accidental as @yidong0635 also said in my patch.
I believe the patch reduces intermittent failures siginificantly.

**Comment by changpe1** [comment]
> Hi @mikeBashStuff @changpe1
> 
> The patch https://review.spdk.io/gerrit/c/spdk/spdk/+/13908 passed CI 3 times successively. I believe this is not accidental as @yidong0635 also said in my patch. I believe the patch reduces intermittent failures siginificantly.

I agree, @mikeBashStuff is your environment that reproduce this issue with this patch same with CI one?

**Comment by mikeBashStuff** [comment]
@changpe1 

I am running this under SPDK without uring support so plain POSIX case. And yes, my environment exactly matches the CI's. I am simply running the test inside a loop with this particular patch being cherry-picked on top - I am hitting failure at different iteration but usually it happens quite fast:

```
 # c=0; while ./test/nvmf/host/timeout.sh --transport=tcp --iso; do ((++c)); done
... 
# echo $c
6
```

**Comment by spdkci** [comment]
Another instance of this failure. Reported by @mikeBashStuff. log: https://ci.spdk.io/public_build/autotest-per-patch_82175.html

**Comment by shuhei-matsumoto** [comment]
@mikeBashStuff 

Thanks for sharing the detailed result. But without this patch, we almost always hit this failure on SPDK per patch CI? How about without this patch? Nothing changed?

**Comment by changpe1** [comment]
> @changpe1
> 
> I am running this under SPDK without uring support so plain POSIX case. And yes, my environment exactly matches the CI's. I am simply running the test inside a loop with this particular patch being cherry-picked on top - I am hitting failure at different iteration but usually it happens quite fast:
> 
> ```
>  # c=0; while ./test/nvmf/host/timeout.sh --transport=tcp --iso; do ((++c)); done
> ... 
> # echo $c
> 6
> ```

By default SPDK will use the POSIX interface, so this is exactly same with the failure CI image.

**Comment by mikeBashStuff** [comment]
@changpe1 @shuhei-matsumoto I scheduled multiple jobs (6, just to make sure I don't block the entire per-patch pipeline) with your patch on the CI to see if we hit this issue there. I will let you know the results when all the jobs are done.

**Comment by mikeBashStuff** [comment]
@changpe1 @shuhei-matsumoto It seems like the patch is indeed helping to mitigate this issue - I run 10 nvmf-tcp-vg jobs with the patch on top and every one of them succeeded. The difference here is that those are 10 separate instances of VM where the test runs only once so I am not sure if this 100% confirms that the issue is gone (as I still can reproduce it on my single VM). In any case, I am going +1 this patch as it definitely has a good impact on the CI.

**Comment by shuhei-matsumoto** [comment]
Thank you so much @mikeBashStuff

**Comment by tomzawadzki** [comment]
The issue was closed via patch merge. Since the issue is still reproduced by @mikeBashStuff, let's keep it open.

**Comment by spdkci** [comment]
Another instance of this failure. Reported by @shuhei-matsumoto. log: https://ci.spdk.io/public_build/autotest-per-patch_82290.html

**Comment by spdkci** [comment]
Another instance of this failure. Reported by @shuhei-matsumoto. log: https://ci.spdk.io/public_build/autotest-per-patch_82289.html

**Comment by shuhei-matsumoto** [comment]
Hi, As my honest intuition matches, this issue has been reproduced even after couple of patches.
I do not have good idea but I will continue my investigation.

**Comment by spdkci** [comment]
Another instance of this failure. Reported by @ksztyber. log: https://ci.spdk.io/public_build/autotest-per-patch_82109.html

**Comment by spdkci** [comment]
Another instance of this failure. Reported by @ksztyber. log: https://ci.spdk.io/public_build/autotest-per-patch_82115.html

**Comment by spdkci** [comment]
Another instance of this failure. Reported by @ksztyber. log: https://ci.spdk.io/public_build/autotest-per-patch_82328.html

**Comment by spdkci** [comment]
Another instance of this failure. Reported by @Comphix. log: https://ci.spdk.io/public_build/autotest-per-patch_82345.html

**Comment by spdkci** [comment]
Another instance of this failure. Reported by @mikeBashStuff. log: https://ci.spdk.io/public_build/autotest-per-patch_82434.html

**Comment by spdkci** [comment]
Another instance of this failure. Reported by @shuhei-matsumoto. log: https://ci.spdk.io/public_build/autotest-per-patch_82362.html

**Comment by spdkci** [comment]
Another instance of this failure. Reported by @shuhei-matsumoto. log: https://ci.spdk.io/public_build/autotest-per-patch_82496.html

**Comment by spdkci** [comment]
Another instance of this failure. Reported by @shuhei-matsumoto. log: https://ci.spdk.io/public_build/autotest-per-patch_82514.html

**Comment by jimharris** [comment]
This is causing a very high rate of failures.  Let's disable this test for now while we continue to debug:

https://review.spdk.io/gerrit/c/spdk/spdk/+/14001

**Comment by spdkci** [comment]
Another instance of this failure. Reported by @benlwalker. log: https://ci.spdk.io/public_build/autotest-per-patch_82491.html

**Comment by spdkci** [comment]
Another instance of this failure. Reported by @benlwalker. log: https://ci.spdk.io/public_build/autotest-per-patch_82485.html

**Comment by spdkci** [comment]
Another instance of this failure. Reported by @benlwalker. log: https://ci.spdk.io/public_build/autotest-per-patch_82486.html

**Comment by spdkci** [comment]
Another instance of this failure. Reported by @benlwalker. log: https://ci.spdk.io/public_build/autotest-per-patch_82660.html

**Comment by spdkci** [comment]
Another instance of this failure. Reported by @jimharris. log: https://ci.spdk.io/public_build/autotest-per-patch_82906.html

**Comment by tomzawadzki** [comment]
Leaving a note here to not misinterpret current state of the issue. The troublesome part of test was disabled https://review.spdk.io/gerrit/c/spdk/spdk/+/14001. This issue should not be closed until the test is re-enabled.

**Comment by tomzawadzki** [comment]
[Bug scrub] @changpe1 have you had a chance to take a look at what it would take to re-enable this test in CI ?

**Comment by spdkci** [comment]
Another instance of this failure. Reported by @AlekseyMarchuk. log: https://ci.spdk.io/public_build/autotest-per-patch_97465.html

**Comment by spdkci** [comment]
Another instance of this failure. Reported by @ksztyber. log: https://ci.spdk.io/public_build/autotest-per-patch_97760.html

**Comment by spdkci** [comment]
Another instance of this failure. Reported by @mgerdts. log: https://ci.spdk.io/public_build/autotest-per-patch_98824.html

**Comment by ksztyber** [comment]
[Bug scrub] Assigning to @jimharris, he'll take a look at what it takes to re-enable the faulty test.

**Comment by jimharris** [comment]
https://review.spdk.io/gerrit/c/spdk/spdk/+/17797

Note @AlekseyMarchuk - I pushed newer version of the #2862 fix, adding that it also fixes #2595 - can you reapply your +2?

**Comment by jimharris** [comment]
Patch merged.
